frontend/src/session/session.jsx

```diff
@@ -0,0 +1,29 @@
+import React, { createContext, useState } from 'react';
+import { getLocalSession, setLocalSession } from '../localStorage';
+
+// Global React.Context value for components to pass into useContext in order to access the session
+const SessionContext = createContext();
+
+// AppContext should wrap the entire application, so that any component in the application
+// can access the global sessions's state as a StoreContext.Consumer
+// eslint-disable-next-line react/prop-types
+const SessionContextProvider = ({ children }) => {
```

Comment:
Our previous discussion was around Application level storage. If you are not planning to use it for anything else but the session, in this case you might not need any of it  (React Context/Provider/Consumer). Below is an example of 3 files (localStorage.js, Login.js, and Authenticated.js)

```js
// localStorage.js
export function setSession(value){
  window.localStorage.setItem('session', JSON.stringify(value));
}

export function getSession(){
  return JSON.parse(window.localStorage.getItem('session'));
}



// login.ts
import { setSession } from './localStorage';

// ...

const tryLogin = async e => {    
  e.preventDefault();
  try {
    const newStore = await api.post('/login', {
      email: e.target.email.value,
      password: e.target.password.value,
    });

    // will add a session to local storage
    setSession(session);    
    
    // will trigger a re-render
    history.push('/dashboard');      
  } catch (error) {
    //
  }
};

// Authenticated.js
import { getSession } from './localStorage';

export default function Authenticated(props) {
  const [authenticated, setAuthenticated ] = React.useState( () => !!getSession()));

  React.useEffect(() => {
    if(!authenticated){
      // navigate a user to the login screen. This will trigger a global re-render
      history.push('/login')
    }
  }, []);


  if (!authenticated) {
     return null;
  }

  return props.children;
}
```

Reply:
@ibeckermayer did some updates to above example

Reply:
@alex-kovoy giving it some thought, I think it is the case that the only piece of fully application level state for this simple of an app is whether or not the user is authenticated. 

OTOH my intention here was to lay down a global state management system that would make the app more maintainable and easier to evolve in the long run, which in my experience is the type of pub-sub system I have here -- any component can "publish" a mutation to the global state, and then each component "subscribes" to pieces of the global state relevant to them, and automatically reacts to changes (basically the Mobx model).

Probably I should make that more clear though by not constraining my naming to be "session"-based, maybe just go back to calling it the more general "store"? Or do you still want me to change it?

Reply:
> maybe just go back to calling it the more general "store"

Yeah, constraining it to a session was a bit confusing. AppStore|AppContext sounds better if you are planing to use it for app level storage.

> a mutation to the global state, and then each component "subscribes" to pieces of the global state relevant to them, and automatically reacts to changes 

Got it. I am familiar with this approach and used it in the past as well although now I prefer hooks with a minimum global state (when it is possible of course)

> Or do you still want me to change it to something closer to your example?

Up to you but if you want to set up a global state I would just name it accordingly.

Reply:
Cool, in that case I'm going to just rename things and stick with what I know for this project (but will dig deeper into the general hooks approach if I manage to land the job 😃)